src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -47,7 +47,8 @@ private string Format(string[] elements)
         var firstElement = true;
         for (var i = 0; i < elements.Length; i++)
         {
-            if (_loggingFields.HasFlag((W3CLoggingFields)(1 << i)))
+            // Custom headers start at index 17
+            if (_loggingFields.HasFlag((W3CLoggingFields)(1 << i)) || i >= 17)
```

Comment:
`|| i >= 17` looks super fragile. We should just pass `string[] additionalHeaders` as a separate array. (e.g. `_w3cLogger.Log(elements, additionalHeaders)`. Or, if that's too inefficient (which I don't think it is relative to everything else), we could pass the index `int startHeaderIndex` (e.g. `_w3cLogger.Log(elements, _fieldsLength)`).

---

FWIW, in `HttpLoggingOptions`, we don't log any of the request headers specified in the `ISet<string>` unless `HttpLoggingFields.RequestHeaders` is set.

https://github.com/dotnet/aspnetcore/blob/8a4b4deb09c04134f22f8d39aae21d212282004f/src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs#L101-L104

It doesn't make much sense to follow that pattern here since `RequestHeaders` isn't its own flag like it is in `HttpLoggingFields` and instead is the combination of the flags for specific request headers.
 https://github.com/dotnet/aspnetcore/blob/8a4b4deb09c04134f22f8d39aae21d212282004f/src/Middleware/HttpLogging/src/W3CLoggingFields.cs#L139
 
And the default value for `W3CLoggerOptions.AdditionalRequestHeaders` is empty unlike `HttpLoggingOptions.RequestHeaders`. I don't think we need to change anything because of this. I'm just noting the similarities and differences so everyone is aware.

Reply:
> || i >= 17 looks super fragile

Agree, I'll rework this